### PR TITLE
Implement source IP RateLimit

### PR DESCRIPTION
**Pull request by zirain**
fixes: https://github.com/envoyproxy/gateway/issues/852

### CODE REVIEW

File: internal/ir/xds.go [hunk 1]
```diff
@@ -694,10 +694,22 @@ type GlobalRateLimit struct {
 type RateLimitRule struct {
 	// HeaderMatches define the match conditions on the request headers for this route.
 	HeaderMatches []*StringMatch
+	// IPMatch define the match conditions on the source IP for this route.
+	IPMatch *IPMatch
```
Comment (arkodg) [inline internal/ir/xds.go:698]: why not list ?

File: internal/xds/translator/ratelimit.go [hunk 2]
```diff
@@ -147,9 +147,26 @@ func buildRouteRateLimits(descriptorPrefix string, global *ir.GlobalRateLimit) [
 			}
 		}
 
+		if rule.IPMatch != nil {
+			// Setup MaskedRemoteAddress action
```
Comment (arkodg) [inline internal/xds/translator/ratelimit.go:151]: from the spec it looks like the masked remote address is being set to RLS https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action-maskedremoteaddress
how are individual IPs getting rate limited in a unique way ?

File: internal/ir/xds.go [hunk 3]
```diff
@@ -694,10 +694,22 @@ type GlobalRateLimit struct {
...
 	Limit *RateLimitValue
 }
 
+type IPMatch struct {
+	CIDR    string
+	IPv6    bool
+	MaskLen int
+}
+
+func (r *RateLimitRule) MatchSetted() bool {
```
Comment (arkodg) [inline internal/ir/xds.go:709]:
```suggestion
func (r *RateLimitRule) IsMatchSet() bool {
```

File: internal/xds/translator/ratelimit.go (hunk 2 above)
Comment (arkodg) [reply internal/xds/translator/ratelimit.go:151]: this boils down to a top level API question, when a user specifies `192.168.1.0/24`and a limit of 2 requests/hourdoes this mean `192.168.1.1` and `192.168.1.2` are treated as the same client so limit is reached when `192.168.1.1` sends 1 request and `192.168.1.2` sends another request within the same hour ? 
Or
is `192.168.1.1` and `192.168.1.2` allowed 2 requests each ?

Comment (zirain) [reply internal/xds/translator/ratelimit.go:151]: in that case you should use `192.168.1.1/32` and `192.168.1.1/32`

File: internal/ir/xds.go (hunk 1 above)
Comment (zirain) [reply internal/ir/xds.go:698]: https://github.com/envoyproxy/gateway/blob/9c4a15053ef0c645591795f4bbf597e2411cc81f/api/v1alpha1/ratelimitfilter_types.go#L105

you want to combine rules here?

File: internal/xds/translator/ratelimit.go (hunk 2 above)
Comment (arkodg) [reply internal/xds/translator/ratelimit.go:151]: thanks for clarifying , we should clarify this in the API https://github.com/envoyproxy/gateway/blob/27d14f8e9c2babe1a7c9f441114bf59b031b1b6d/api/v1alpha1/ratelimitfilter_types.go#L101 i.e. `All IP Addresses within the specified CIDR are treated as a single client selector and share the rate limit value`

In the future, we could add a `Distinct` match type so each IP is treated in a unique way

File: internal/ir/xds.go [hunk 4]
```diff
@@ -694,10 +694,22 @@ type GlobalRateLimit struct {
 type RateLimitRule struct {
 	// HeaderMatches define the match conditions on the request headers for this route.
 	HeaderMatches []*StringMatch
+	// IPMatch define the match conditions on the source IP for this route.
+	IPMatch *IPMatch
 	// Limit holds the rate limit values.
 	Limit *RateLimitValue
 }
 
+type IPMatch struct {
```
Comment (arkodg) [inline internal/ir/xds.go:703]: nit: `CIDRMatch` ?

File: internal/ir/xds.go (hunk 1 above)
Comment (arkodg) [reply internal/ir/xds.go:698]: I think this is non needed, my bad

File: internal/xds/translator/ratelimit.go (hunk 2 above)
Comment (zirain) [reply internal/xds/translator/ratelimit.go:151]: `MaskedRemoteAddress` is usefull for both Global and Local ratelimit.
In istio, we want to user local ratelimit between pod and pod, pod ip is short live, `marked_remote_address` will be very useful in this scenario.

**Comment by arkodg** [comment]
hey @zirain has this been tested end to end ? would be great to track it with a GH issue so we track docs completion for this feature

File: internal/xds/translator/testdata/in/ratelimit-config/masked-remote-address-match.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+- "*"
+routes:
+- name: "first-route"
+  rateLimit:
+    global:
+      rules:
+      - cidrMatch:
+          cidr: "192.168.0.0/16"
+          ipv6: false
+          ones: 16
```
Comment (arkodg) [inline internal/xds/translator/testdata/in/ratelimit-config/masked-remote-address-match.yaml:14]: ones ?

Comment (zirain) [reply internal/xds/translator/testdata/in/ratelimit-config/masked-remote-address-match.yaml:14]: come from:
```
// Size returns the number of leading ones and total bits in the mask.
// If the mask is not in the canonical form--ones followed by zeros--then
// Size returns 0, 0.
func (m IPMask) Size() (ones, bits int) {
	ones, bits = simpleMaskLength(m), len(m)*8
	if ones == -1 {
		return 0, 0
	}
	return
}
```

Comment (zirain) [reply internal/xds/translator/testdata/in/ratelimit-config/masked-remote-address-match.yaml:14]: maybe change it to `maskLen`?

**Comment by zirain** [comment]
open https://github.com/envoyproxy/gateway/issues/1144

File: internal/xds/translator/testdata/in/ratelimit-config/masked-remote-address-match.yaml (hunk 5 above)
Comment (arkodg) [reply internal/xds/translator/testdata/in/ratelimit-config/masked-remote-address-match.yaml:14]: yeah, isn't this the struct
```
type CIDRMatch struct {
 	CIDR    string
 	IPv6    bool
 	MaskLen int
 }
 ```
 so the input should be `maskLen`
 
 rateLimitConfig doesnt use `maskLen` so this test is passing :)

Comment (zirain) [reply internal/xds/translator/testdata/in/ratelimit-config/masked-remote-address-match.yaml]: yeah, I found it.

**Review by arkodg — APPROVED** [review]
LGTM, thanks for building out this useful feature !

File: internal/xds/translator/ratelimit.go (hunk 2 above)
Comment (arkodg) [reply internal/xds/translator/ratelimit.go:151]: raised https://github.com/envoyproxy/gateway/pull/1145 to clarify docs

**Review by Xunzhuo — APPROVED** [review]
LGTM, thanks! Can you update the ratelimit doc to describe this?

**Pull request merged**